src/bc_protocol/time.rs

```diff
@@ -0,0 +1,132 @@
+use super::{BcCamera, Error, Result, RX_TIMEOUT};
+use crate::bc::{model::*, xml::*};
+use time::{date, Date, OffsetDateTime, PrimitiveDateTime, Time, UtcOffset};
+
+impl BcCamera {
+    pub fn get_time(&self) -> Result<Option<OffsetDateTime>> {
+        let connection = self
+            .connection
+            .as_ref()
+            .expect("Must be connected to get time");
+        let sub_get_general = connection.subscribe(MSG_ID_GET_GENERAL)?;
+        let get = Bc {
+            meta: BcMeta {
+                msg_id: MSG_ID_GET_GENERAL,
+                client_idx: 0,
+                encrypted: true,
+                class: 0x6414,
+            },
+            body: BcBody::ModernMsg(ModernMsg::default()),
+        };
+
+        sub_get_general.send(get)?;
+        let msg = sub_get_general.rx.recv_timeout(RX_TIMEOUT)?;
+
+        if let BcBody::ModernMsg(ModernMsg {
+            xml:
+                Some(BcXml {
+                    system_general:
+                        Some(SystemGeneral {
+                            time_zone: Some(time_zone),
+                            year: Some(year),
+                            month: Some(month),
+                            day: Some(day),
+                            hour: Some(hour),
+                            minute: Some(minute),
+                            second: Some(second),
+                            ..
+                        }),
+                    ..
+                }),
+            ..
+        }) = msg.body
+        {
+            let datetime = match try_build_timestamp(
+                time_zone, year, month, day, hour, minute, second
+            ) {
+                Ok(dt) => dt,
+                Err(e) => return Err(Error::UnintelligibleReply {
+                    reply: msg,
+                    why: "Could not parse date",
+                })
+            };
+
+            // This code was written in 2020; I'm trying to catch all the possible epochs that
+            // cameras might reset themselves to. My B800 resets to Jan 1, 1999, but I can't
+            // guarantee that Reolink won't pick some newer date.  Therefore, last year ought
+            // to be new enough, yet still distant enough that it won't interfere with anything
+            const BOUNDARY: Date = date!(2019-01-01);
```

Comment:
How far out of sync does reolink cameras let us be before we get dissconnects? It may be sufficient to just have this as NOW-10mins I assume that anyone running a camera would like the correct time stamps on there videos.

Reply:
Hm that's a good point. Maybe I should take that approach and let someone yell if they don't like it.

Reply:
So I am actually going to leave this particular snippet the way it is, because it's detecting time-not-set-at-all, not time-wrong.  It should be the job of the caller of this function to detect time-wrong.

Reply:
I don't dispute that fixing time-wrong would be nice.